### PR TITLE
Allocate Pods in Custom VPC - Vanilla Kubernetes

### PR DESCRIPTION
**Pull request by phudtran**
What does this change do?
This change allows for pods to be allocated into user specified VPCs and Subnets via annotations.

Why is it needed?
Needed for users to fully utilizes the capabilities of Mizar in Kubernetes

How was this tested?
Python functional tests
- Same net communication in created custom VPC and Subnet.
- Cross net communication in create custom VPC and Subnet.
- Pod allocated in default subnet if no Subnet specified by user for VPC.

### CODE REVIEW

File: teste2e/common/k8s.py
Comment (pdgetrf) [inline teste2e/common/k8s.py]: maybe put "Provisioned" in a variable to prevent duplication (also seen on line #65)

File: cli/mizarapi.py
Comment (pdgetrf) [inline cli/mizarapi.py:29]: where's the vni parameter used?

**Comment by pdgetrf** [comment]
Looks good to me. In testing cross VPC test should also be included if not yet already.

Also please add @jshaofuturewei and @chenqianfzh as "curiosity" approver.

**Review by pdgetrf — APPROVED** [review]
LGTM

File: mizar/obj/tests/test_net1.yaml
Comment (jshaofuturewei) [inline mizar/obj/tests/test_net1.yaml:28]: Not sure if the vpc whose vni is 15654902 DO exist when we apply test-net1.yaml

**Review by jshaofuturewei — APPROVED** [review]

**Comment by jshaofuturewei** [comment]
I tried to start clusters with the request using the command "kind-setup.sh dev" but got the following error.  Could you try locally  to see if it works? Thanks!

[2021-08-25 20:53:44,155] luigi-interface      [ERROR   ] [pid 8] Worker Worker(salt=229995786, workers=1, host=kind-worker2, username=root, pid=8) failed    k8sPodCreate(param=<mizar.common.wf_param.HandlerParam object at 0x7f0ff1f85d00>)
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/luigi/worker.py", line 199, in run
    new_deps = self._run_get_new_deps()
  File "/usr/local/lib/python3.8/site-packages/luigi/worker.py", line 141, in _run_get_new_deps
    task_gen = self.task.run()
  File "/usr/local/lib/python3.8/site-packages/mizar/dp/mizar/workflows/builtins/pods/create.py", line 99, in run
    self.raise_temporary_error("Droplet not yet created.")
  File "/usr/local/lib/python3.8/site-packages/mizar/common/workflow.py", line 46, in raise_temporary_error
    raise Exception(self.error_message)
Exception: Droplet not yet created.
[2021-08-25 20:53:44,163] luigi-interface      [INFO    ] Informed scheduler that task   k8sPodCreate__mizar_common_wf_02f7e369df   has status   FAILED
[2021-08-25 20:53:44,167] luigi-interface      [INFO    ] Worker Worker(salt=229995786, workers=1, host=kind-worker2, username=root, pid=8) was stopped. Shutting down Keep-Alive thread
[2021-08-25 20:53:44,168] luigi-interface      [INFO    ]
===== Luigi Execution Summary =====

Scheduled 1 tasks of which:
* 1 failed:
    - 1 k8sPodCreate(param=<mizar.common.wf_param.HandlerParam object at 0x7f0ff1f85d00>)

File: mizar/dp/mizar/workflows/builtins/pods/create.py
Comment (vinaykul) [inline mizar/dp/mizar/workflows/builtins/pods/create.py]: Please make mizar.com/vpc a constant. 
e.g: https://github.com/CentaurusInfra/mizar/pull/528/files#diff-24aa2233e4c166c6d00c69676f1cac396e7a1b68838af6de30f66318fdf9bbbcL104

File: mizar/dp/mizar/workflows/builtins/pods/create.py
Comment (vinaykul) [inline mizar/dp/mizar/workflows/builtins/pods/create.py:75]: This as well.

**Review by vinaykul — APPROVED** [review]
Just a couple of comments.
LGTM.

File: mizar/dp/mizar/workflows/builtins/pods/create.py
Comment (phudtran) [reply mizar/dp/mizar/workflows/builtins/pods/create.py:75]: Fixed

File: mizar/obj/tests/test_net1.yaml
Comment (phudtran) [reply mizar/obj/tests/test_net1.yaml:28]: This was just for some manual testing.

File: teste2e/common/k8s.py
Comment (phudtran) [reply teste2e/common/k8s.py]: fixed thanks for the catch!

**Comment by pdgetrf** [comment]
@phudtran I want to try this out but in a "real" K8s cluster with actual nodes.

Does Mizar have a document and/or script for installation Mizar into a k8s cluster with actual nodes instead of not KIND nodes?

**Comment by phudtran** [comment]
> @phudtran I want to try this out but in a "real" K8s cluster with actual nodes.
> 
> Does Mizar have a document and/or script for installation Mizar into a k8s cluster with actual nodes instead of not KIND nodes?

@pdgetrf 
Unfortunately, there is no easy way to try this out on a real Kubernetes cluster until the PR is merged and the dockerhub images have been updated.

For now, you would have to pull this branch and build the images required locally on each of your machines.
[An example here which is done for our CI tests](https://github.com/CentaurusInfra/mizar/blob/dev-next/.github/workflows/e2e_tests.yaml#L94)

Once the PR is merged, and the images have been pushed and you can simply use this [yaml](https://github.com/CentaurusInfra/mizar/blob/dev-next/etc/deploy/deploy.mizar.yaml) to install mizar.

File: mizar/dp/mizar/workflows/builtins/pods/create.py
Comment (phudtran) [reply mizar/dp/mizar/workflows/builtins/pods/create.py]: Fixed

**Comment by pdgetrf** [comment]
> > @phudtran I want to try this out but in a "real" K8s cluster with actual nodes.
> > Does Mizar have a document and/or script for installation Mizar into a k8s cluster with actual nodes instead of not KIND nodes?
> 
> @pdgetrf
> Unfortunately, there is no easy way to try this out on a real Kubernetes cluster until the PR is merged and the dockerhub images have been updated.
> 
> For now, you would have to pull this branch and build the images required locally on each of your machines.
> [An example here which is done for our CI tests](https://github.com/CentaurusInfra/mizar/blob/dev-next/.github/workflows/e2e_tests.yaml#L94)
> 
> Once the PR is merged, and the images have been pushed and you can simply use this [yaml](https://github.com/CentaurusInfra/mizar/blob/dev-next/etc/deploy/deploy.mizar.yaml) to install mizar.

this sounds good. I'll wait for the PR to merge then. thanks.

**Comment by phudtran** [comment]
> I tried to start clusters with the request using the command "kind-setup.sh dev" but got the following error. Could you try locally to see if it works? Thanks!
> 
> [2021-08-25 20:53:44,155] luigi-interface [ERROR ] [pid 8] Worker Worker(salt=229995786, workers=1, host=kind-worker2, username=root, pid=8) failed k8sPodCreate(param=<mizar.common.wf_param.HandlerParam object at 0x7f0ff1f85d00>)
> Traceback (most recent call last):
> File "/usr/local/lib/python3.8/site-packages/luigi/worker.py", line 199, in run
> new_deps = self._run_get_new_deps()
> File "/usr/local/lib/python3.8/site-packages/luigi/worker.py", line 141, in _run_get_new_deps
> task_gen = self.task.run()
> File "/usr/local/lib/python3.8/site-packages/mizar/dp/mizar/workflows/builtins/pods/create.py", line 99, in run
> self.raise_temporary_error("Droplet not yet created.")
> File "/usr/local/lib/python3.8/site-packages/mizar/common/workflow.py", line 46, in raise_temporary_error
> raise Exception(self.error_message)
> Exception: Droplet not yet created.
> [2021-08-25 20:53:44,163] luigi-interface [INFO ] Informed scheduler that task k8sPodCreate__mizar_common_wf_02f7e369df has status FAILED
> [2021-08-25 20:53:44,167] luigi-interface [INFO ] Worker Worker(salt=229995786, workers=1, host=kind-worker2, username=root, pid=8) was stopped. Shutting down Keep-Alive thread
> [2021-08-25 20:53:44,168] luigi-interface [INFO ]
> ===== Luigi Execution Summary =====
> 
> Scheduled 1 tasks of which:
> 
> * 1 failed:
>   
>   * 1 k8sPodCreate(param=<mizar.common.wf_param.HandlerParam object at 0x7f0ff1f85d00>)

You may have have stale images.
Try running these commands and then kind-setup again.
```
docker stop $(docker ps -aq) && docker system prune -f
docker rmi -f $(docker images -a -q)
docker volume rm $(docker volume ls -qf dangling=true)
```

**Comment by pdgetrf** [comment]
@phudtran are we still waiting for more approval for this PR?

**Comment by phudtran** [comment]
> @phudtran are we still waiting for more approval for this PR?

Just need @vinaykul to merge it :)

**Comment by jshaofuturewei** [comment]
> @phudtran are we still waiting for more approval for this PR?

It works now. Thanks for the updates.

**Comment by pdgetrf** [comment]
> > @phudtran are we still waiting for more approval for this PR?
> 
> Just need @vinaykul to merge it :)

Hi @vinaykul 

Could you please merge this if all is cleared before your vacation? Next week we are hoping to test and use the feature from this PR. Thanks.

**Comment by pdgetrf** [comment]
> > @phudtran are we still waiting for more approval for this PR?
> 
> Just need @vinaykul to merge it :)

Seems @vinaykul is on vacation now. Anyone else can merge this?

File: mizar/obj/tests/test_net1.yaml
Comment (pdgetrf) [reply mizar/obj/tests/test_net1.yaml:28]: I have a similar question here. if the vpc1 that's already created in the cluster has vni "3456", do we put "3456" here? what does this vni do?

Comment (pdgetrf) [reply mizar/obj/tests/test_net1.yaml:28]: and it goes back to the original question, which is the unique identifier of a vpc? is it name or vni. both options are provided here. what happens if the provided vni here is different from the actual vni of the vpc1? 

I tried to create a subnet with vpc1 as name and "wrong" vni, it was created:

image

But now it gets confusing. Is this vpc1 with vni 15654902 the same as the "vpc1" with vni 10731071?

**Pull request merged**
